diff -ruN a/src/install/PackageInstall.zig b/src/install/PackageInstall.zig
--- a/src/install/PackageInstall.zig	2026-07-10 23:00:04.144813438 +0800
+++ b/src/install/PackageInstall.zig	2026-07-10 23:00:04.340813750 +0800
@@ -372,7 +372,7 @@
     pub var supported_method: Method = if (Environment.isMac)
         Method.clonefile
     else
-        Method.hardlink;
+        Method.symlink; // hmdfs does not support hardlink
 
     fn installWithClonefileEachDir(this: *@This(), destination_dir: std.fs.Dir) !Result {
         var cached_package_dir = bun.openDir(this.cache_dir, this.cache_dir_subpath) catch |err| return Result.fail(err, .opening_cache_dir, @errorReturnTrace());
diff -ruN a/src/resolver/resolver.zig b/src/resolver/resolver.zig
--- a/src/resolver/resolver.zig	2026-07-10 23:00:04.196815469 +0800
+++ b/src/resolver/resolver.zig	2026-07-11 00:11:18.309569844 +0800
@@ -4114,7 +4114,7 @@
             info.package_json_for_dependencies = info.package_json_for_dependencies orelse parent_.package_json_for_dependencies;
 
             // Make sure "absRealPath" is the real path of the directory (resolving any symlinks)
-            if (!r.opts.preserve_symlinks) {
+            if (!r.opts.preserve_symlinks and !parent_.isNodeModules() and !parent_.isInsideNodeModules()) { // hmdfs: symlinks in node_modules point to cache
                 if (parent_.getEntries(r.generation)) |parent_entries| {
                     if (parent_entries.get(base)) |lookup| {
                         if (entries.fd.isValid() and !lookup.entry.cache.fd.isValid() and r.store_fd) lookup.entry.cache.fd = entries.fd;
